mesonbuild/scripts/meson_exe.py

```diff
@@ -64,14 +65,30 @@ def run_exe(exe):
         child_env['PATH'] = (os.pathsep.join(exe.extra_paths + ['']) +
                              child_env['PATH'])
         if exe.exe_runner and mesonlib.substring_is_in_list('wine', exe.exe_runner.get_command()):
-            wine_paths = ['Z:' + p for p in exe.extra_paths]
-            wine_path = ';'.join(wine_paths)
-            # Don't accidentally end with an `;` because that will add the
-            # current directory and might cause unexpected behaviour
-            if 'WINEPATH' in child_env:
-                child_env['WINEPATH'] = wine_path + ';' + child_env['WINEPATH']
-            else:
-                child_env['WINEPATH'] = wine_path
+            seen = set()
+            wine_paths = ['Z:' + p for p in exe.extra_paths if not (p in seen or seen.add(p))]
+            wine_paths += [p for p in child_env.get('WINEPATH', '').split(';') if not (p in seen or seen.add(p))]
+
+            getShortPathScript = '%s.bat' % str(uuid.uuid4()).lower()[:5]
+            with open(getShortPathScript, mode='w') as f:
+                f.write("@ECHO OFF\nfor %%x in (%*) do (\n echo|set /p=;%~sx\n)\n")
+                f.flush()
+            try:
+                with open(os.devnull, 'w') as stderr:
+                    wine_path = subprocess.check_output(
+                        exe.exe_runner.get_command() +
+                        ['cmd', '/C', getShortPathScript] + wine_paths,
+                        stderr=stderr).decode('utf-8')
+            except subprocess.CalledProcessError as e:
+                print("Could not get short paths: %s" % e)
+                wine_path = ';'.join(wine_paths)
+            finally:
+                os.remove(getShortPathScript)
+            if len(wine_path) > 2048:
+                raise AssertionError('WINEPATH size {} > 2048'
+                                     ' this will cause random failure.'.format(
+                                         len(wine_path)))
```

Comment:
Use `MesonException` or its derived class instead. Plain exceptions cause the backtrace to be printed.